src/lib/ext-monitor.js

```diff
@@ -0,0 +1,91 @@
+export default class ExtensionMonitor {
+  logs = [];
+  extensionMapList = new Map([]);
+
+  async isExtensionPageOpen() {
+    const tab = await browser.tabs.query({
+      url: browser.runtime.getURL('activitylog/activitylog.html'),
+    });
+    if (tab.length) return true;
+    return false;
+  }
+  // If the extension page is open, each logs will be send to
+  // extension page (as soon as it is encountered), if it is open.
+  async sendLogs(details) {
+    const isExtPageOpen = await this.isExtensionPageOpen();
+    if (isExtPageOpen) {
+      await browser.runtime.sendMessage({
+        updateLogs: details,
+      });
+    }
+  }
+
+  logListener() {
+    const listener = async (details) => {
+      // Everytime a log is encountered,
+      // it is being pushed to logs array.
+      this.logs.push(details);
+      //it checks if the extension page is open so that it can send log
+      await this.sendLogs(details);
+    };
+    return listener;
+  }
+
+  setExtensions(extensionsArray) {
+    extensionsArray.forEach((extension) => {
+      this.startMonitor(extension.id);
+    });
+    return Promise.resolve('ext-monitor-started');
+  }
+
+  startMonitor(extensionId) {
+    if (this.extensionMapList.has(extensionId)) {
```

Comment:
I would expect startMonitor to early exit also when extensionId is the id of the extension activity monitor extension ([onExtensionActivity.addListener is also preventing an extension to monitor itself internally, but it also log an error](https://searchfox.org/mozilla-central/rev/1304678d837c17811617754fd446794b774afe94/toolkit/components/extensions/parent/ext-activityLog.js#27-29), which could be misleading if we are investigating a bug in the extension, and so it would be reasonable to prevent the extension from calling onExtensionActivity.addListener with its own extension id in the first place).

Reply:
Done. I have excluded monitoring our own extension. :-)